File: 416.分割等和子集.c
/*
 * @lc app=leetcode.cn id=416 lang=c
 *
 * [416] 分割等和子集
 */

// @lc code=start
bool canPartition(int *nums, int numsSize)
{
  int *dp, i, j, sum = 0;
  for (i = 0; i < numsSize; i++)
    sum += nums[i];
  if (sum % 2 != 0)
    return false;
  else
  {
    dp = (int *)malloc(sizeof(int) * (sum / 2 + 1));
    for (i = 0; i < sum / 2 + 1; i++)
      dp[i] = 0;

    for (i = 0; i < numsSize; i++)
    {
      for (j = sum / 2; j >= nums[i]; j--)
      {
        dp[j] = dp[j] > dp[j - nums[i]] + nums[i] ? dp[j] : dp[j - nums[i]] + nums[i];
      }
    }
    if (dp[sum / 2] == sum / 2)
      return true;
    else
      return false;
  }
}
// @lc code=end

File: store/416.分割等和子集.js
/*
 * @lc app=leetcode.cn id=416 lang=javascript
 *
 * [416] 分割等和子集
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canPartition = function (nums) {
  // 法一
  // let sum, dp;
  // sum = nums.reduce((pre, next) => pre + next);
  // if (sum % 2) return false;
  // else sum /= 2;
  // dp = new Array(nums.length).fill(0).map(() => Array(sum + 1).fill(0));

  // for (let j = 1; j <= sum; j++) {
  //   if (nums[0] <= j) dp[0][j] = nums[0];
  // }

  // for (let i = 1; i < nums.length; i++) {
  //   for (let j = 0; j <= sum; j++) {
  //     if (j < nums[i]) dp[i][j] = dp[i - 1][j];
  //     else dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - nums[i]] + nums[i]);
  //   }
  // }
  // return dp[nums.length - 1][sum] === sum;

  // 法二
  let sum, dp;
  sum = nums.reduce((pre, cur) => pre + cur);
  if (sum % 2) return false;
  else sum /= 2;
  dp = new Array(sum + 1).fill(0);
  // dp[0] = 0

  for (let i = 0; i < nums.length; i++) {
    for (let j = sum; j > -1; j--) {
      if (j >= nums[i]) dp[j] = Math.max(dp[j], dp[j - nums[i]] + nums[i]);
    }
  }

  return dp[sum] === sum;
};
// @lc code=end
